### ISSUE TITLE
Basic analysis for `accumulate` exercise

### ISSUE THREAD

**Issue by ThiagosLima** [exercism/elixir-analyzer#67]
Hi, as commented in the [Elixir v3 repo #560](https://github.com/exercism/elixir/issues/560), maybe the `accumulate` exercise should have an analyse for `collect`, `map` and `fmap`.

**Comment by neenjaw** [comment]
Thanks for logging this request here! I think we could get to it soon. If you are interested in tackling it, you are more than welcome to, could set aside some time to walk you through it this weekend.

**Comment by angelikatyborska** [comment]
👍 

As the instruction for `accumulate` say (https://github.com/exercism/elixir/blob/d18ac65f0302c4f6b2ba6ab0184387b38cc8dadd/exercises/practice/accumulate/.docs/instructions.md#L24-L26)
> Keep your hands off that collect/map/fmap/whatchamacallit functionality
> provided by your standard library!
> Solve this one yourself using other basic tools instead.

To translate that into Elixir-specific terms: I think we should forbid the whole Enum, List, and Stream modules, and also list comprehensions. Just like we did for `bird-count`, and just like we want to do for `flatten-array` (https://github.com/exercism/elixir-analyzer/issues/60). In fact, I think the analyzer for `accumulate` and `flatten-array` could be done at the same time and could be virtually identical :)

**Comment by ThiagosLima** [comment]
I know just the basics of Elixir but I believe this is a good opportunity to learn a little more. Can I try to do the issue?

**Comment by neenjaw** [comment]
Yea, for sure, I'll write out a bit of a guide of the things to need to do to get it going.  Should be able to have it posted tonight

**Comment by neenjaw** [comment]
## Things to be done

- [ ] Use `bird-count`'s [analyzer test suite](https://github.com/exercism/elixir-analyzer/blob/main/lib/elixir_analyzer/test_suite/bird_count.ex) as a basis for your work
- [ ] Update [`config.exs`](https://github.com/exercism/elixir-analyzer/blob/main/config/config.exs) with the added analyzer test
- [ ] Create a test for it, can follow the example for [bird-count in the test dir](https://github.com/exercism/elixir-analyzer/blob/main/test/elixir_analyzer/test_suite/bird_count_test.exs

Once that is done, `mix test` should be able to run your analyzer

We would also need a PR to `website-copy` with a accumulate-specific version of https://github.com/exercism/website-copy/blob/main/analyzer-comments/elixir/bird-count/use_recursion.md

---

All in all, not too much elixir programming, but it should give you an idea about the using elixir in a project, using configs etc.  We have implemented a lot of custom constructs with metaprogramming to hide away most of the working bits, and make these files reasonably declarative

**Comment by neenjaw** [comment]
Just shout if you need any help, @angelikatyborska and I are around!

**Comment by ThiagosLima** [comment]
Hmmm I have a question. As @angelikatyborska said before, the code for `Accumulate`, `Flatten Array` and `Bird Count` will be almost the same. Should we keep all of them in separate files?

**Comment by angelikatyborska** [comment]
@ThiagosLima Yes, let's keep them in separate files even if they will be word for word the same. We decided on a convention that each exercise has its own analysis module. Later, we might need to add more detailed analysis for one of those exercises and then it will be useful that it's separate 🙂 

If we notice that some patterns repeat a lot, we'll think about extracting them and reusing somehow later 🤔

**Comment by ThiagosLima** [comment]
Hi, guys! I am almost finished the issue. I took some time to study macros and tests to understand the analyzer better. Sorry for take so long :sweat_smile:.

Firstly, the project is really nice and well documented. Congratulations! You guys are awesome! :blush:

I am taking the [bird count](https://github.com/exercism/elixir-analyzer/blob/main/test/elixir_analyzer/test_suite/bird_count_test.exs) as an exemple. But I have some questions about the tests:

1 - Isn't the line 52 already test in the 31-33 test? I was wondering if I have to add a line like that in this issue.
2 - Is it necessary to test many cases of list comprehension? If I do something like this would take all of comprehension cases?
```elixir
defmodule Accumulate do
	def accumulate(_list, _fun) do
		for l <- [], do: l
	end
end
```
3 - Do I have to write the solution for test functions used in the exercise to test the modules? I tryed to do that but I was wondering that maybe write the smallest function to test could be better since we are just seeking for the module. Ex:

```elixir
defmodule Accumulate do
	def accumulate(list, fun), do: List.foldr(list, [], fn x, acc -> [fun.(x) | acc] end)
end
```
```elixir
defmodule Accumulate do
	def accumulate(_list, _fun), do: List.first([])
end

**Comment by angelikatyborska** [comment]
Don't worry about it! We really appreciate that you're working on this issue 🙏 take your time.

> 1 - Isn't the line 52 already test in the 31-33 test? I was wondering if I have to add a line like that in this issue.

You're totally right. I made a mistake. At line 52, it was supposed to be a different usage of the alias. Something like this: https://github.com/exercism/elixir-analyzer/pull/74

> 2 - Is it necessary to test many cases of list comprehension?

No, not really :) I did that initially because I wasn't sure if the bird-count analyzer is finding them correctly. If you write a rule like this one, with `assert_no_call`, a single test case for it will be enough: https://github.com/exercism/elixir-analyzer/blob/main/lib/elixir_analyzer/test_suite/bird_count.ex#L80-L84

> 3 - Do I have to write the solution for test functions used in the exercise to test the modules? 

It would be good if the test snippets were related to a real solution somehow, but they don't have to form a full solution that would pass the test suite of the exercise. The snippets that you're suggesting look good :). The more related to a real solution the snippet is, the easier it will be to remember why we're even writing an analyzer for such a case. So I would suggest avoiding writing functions named `foo` and `bar` in the test snippets and it will be fine 👍.
